### PR TITLE
Display recommended match users in frontpage widget

### PR DESCRIPTION
**Pull request by lagerros**
* Refactor row component to be able to add more rows
* Write new query for fetching recommended users. Basically just your upvoted users, but filtered for ones who have written two comments on your posts or replied to three of your comments

image

### CODE REVIEW

File: packages/lesswrong/server/resolvers/userResolvers.ts [hunk 1]
```diff
@@ -535,3 +535,18 @@ defineQuery({
...
+  name: "GetDialogueRecommendedUsers",
+  resultType: "[User]!",
+  fn: async (root, _, context) => {
+    const { currentUser } = context
+    if (!currentUser) {
+      throw new Error('User must be logged in to get recommended users');
+    }
+
+    const upvotedUsers = await new UsersRepo().getUsersTopUpvotedUsers(currentUser, 30, 720)
+    const recommendedUsers = await new UsersRepo().getDialogueRecommendedUsers(currentUser._id, upvotedUsers);
```
Comment (oetherington) [inline packages/lesswrong/server/resolvers/userResolvers.ts:549]:
```suggestion
    const upvotedUsers = await context.repos.users.getUsersTopUpvotedUsers(currentUser, 30, 720)
    const recommendedUsers = await context.repos.users.getDialogueRecommendedUsers(currentUser._id, upvotedUsers);
```

File: packages/lesswrong/server/repos/UsersRepo.ts [hunk 2]
```diff
@@ -400,6 +399,36 @@ export default class UsersRepo extends AbstractRepo<DbUser> {
...
+  const upvotedUserIds = upvotedUsers.map(user => user._id);
+
+  return this.any(`
+    SELECT u.*
+    FROM unnest($2::text[]) AS uv(_id)
+    INNER JOIN "Users" AS u ON uv._id = u._id
+    WHERE
+      -- Don't recommend you users who've never commented twice your posts OR replied 3 times to your comments 
+      EXISTS (
+        SELECT 2
```
Comment (oetherington) [inline packages/lesswrong/server/repos/UsersRepo.ts:412]: This doesn't do what you think it does - it ignores the table you're selecting from and selects a single unnamed row containing the literal value "2". This always exists, so the `EXISTS` clause always returns true. You need to use `COUNT` like below.

File: packages/lesswrong/server/repos/UsersRepo.ts [hunk 3]
```diff
@@ -400,6 +399,36 @@ export default class UsersRepo extends AbstractRepo<DbUser> {
     `, [userId]);
   }
 
+  async getDialogueRecommendedUsers(userId: string, upvotedUsers:UpvotedUser[], limit = 100): Promise<DbUser[]> {
+  const upvotedUserIds = upvotedUsers.map(user => user._id);
```
Comment (oetherington) [inline packages/lesswrong/server/repos/UsersRepo.ts:403]: The entire contents of this function needs to be indented.

**Review by oetherington — COMMENTED** [review]
I've left a few comments, otherwise no objections from EAF

**Comment by lagerros** [comment]
this will get incorporated in PR #8322 instead, so closing

**Pull request closed without merging**